pandas/io/common.py

```diff
@@ -584,48 +602,57 @@ def get_handle(
             msg = f"Unrecognized compression type: {compression}"
             raise ValueError(msg)
 
+        assert not isinstance(f, str)
         handles.append(f)
 
     elif is_path:
         # Check whether the filename is to be opened in binary mode.
         # Binary mode does not support 'encoding' and 'newline'.
         is_binary_mode = "b" in mode
-
+        assert isinstance(path_or_buf, str)
         if encoding and not is_binary_mode:
             # Encoding
             f = open(path_or_buf, mode, encoding=encoding, errors=errors, newline="")
-        elif is_text and not is_binary_mode:
-            # No explicit encoding
-            f = open(path_or_buf, mode, errors="replace", newline="")
         else:
             # Binary mode
             f = open(path_or_buf, mode)
         handles.append(f)
 
     # Convert BytesIO or file objects passed with an encoding
-    if is_text and (compression or isinstance(f, need_text_wrapping)):
-        from io import TextIOWrapper
-
-        g = TextIOWrapper(f, encoding=encoding, errors=errors, newline="")
-        if not isinstance(f, (BufferedIOBase, RawIOBase)):
-            handles.append(g)
-        f = g
+    is_wrapped = False
+    if is_text and (
+        compression
+        or isinstance(f, need_text_wrapping)
+        or "b" in getattr(f, "mode", "")
+    ):
+        f = TextIOWrapper(
+            f, encoding=encoding, errors=errors, newline=""  # type: ignore[arg-type]
+        )
+        handles.append(f)
+        is_wrapped = True
 
     if memory_map and hasattr(f, "fileno"):
+        assert not isinstance(f, str)
         try:
-            wrapped = _MMapWrapper(f)
+            wrapped = _MMapWrapper(f)  # type: ignore[arg-type]
```

Comment:
can you cast or type wrapped before here? (its a union?)

Reply:
I cast it to `mmap.mmap` and added it to the `Buffer` union.